pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java

```diff
@@ -97,10 +122,38 @@ public BaseDataBlock getDataBlock() {
     return _dataBlock;
   }
 
+  /**
+   * Return the type of block (one of ROW, COLUMNAR, or METADATA).
+   *
+   * @return return type of block
+   */
   public BaseDataBlock.Type getType() {
     return _type;
   }
 
+  /**
+   * Return whether a transferable block is at the end of a stream.
+   *
+   * <p>End of stream is different from data block with 0-rows. which can indicate that one partition of the execution
+   * returns no rows. but that doesn't mean the rest of the partitions are also finished.
+   * <p>When an exception is caught within a stream, no matter how many outstanding data is pending to be received,
+   * it is considered end of stream because the exception should bubble up immediately.
+   *
+   * @return whether this block is the end of stream.
+   */
+  public boolean isEndOfStreamBlock() {
+    return _isErrorBlock || _type == BaseDataBlock.Type.METADATA;
```

Comment:
This condition is not very intuitive.

- In both cases (eos or error), the block is a metadata block
- The block should be self-descriptive and tell if the metadata block is eos or error.
- If in future, we leverage METADATA block in some other way it may not necessarily imply EOS block ?

Reply:
good idea. let me think about this.

Reply:
oh i remembered why this was done like this --> this was for leaf node. since the data block was a row-based data block with 0-rows, and an exception map not empty.. so I had to make the change. 

now that I think of it I should rewrap and return at this stage. so that I don't have to rewrap again in the future.